extension/observer/endpointswatcher_test.go

```diff
@@ -20,76 +20,173 @@ import (
+	lister, watcher, notifyOne := setup(t)
+	defer watcher.StopListAndWatch()
+
+	lister.addEndpoint(0)
+	notifyOne.numToBeAdded = 1
+	watcher.ListAndWatch(notifyOne)
+
+	notifyTwo := newMockNotififier(t, "notify2")
+	lister.addEndpoint(1)
+	notifyTwo.numToBeAdded = 2
+	watcher.ListAndWatch(notifyTwo)
+
+	expectedOne := map[EndpointID]Endpoint{"0": {ID: "0"}}
+	require.Equal(t, expectedOne, existingEndpoints(t, watcher, notifyOne.ID()))
+
+	expectedTwo := map[EndpointID]Endpoint{"0": {ID: "0"}, "1": {ID: "1"}}
+	require.Equal(t, expectedTwo, existingEndpoints(t, watcher, notifyTwo.ID()))
+
+	lister.addEndpoint(2)
+	lister.addEndpoint(3)
+	notifyOne.numToBeAdded = 3
+	notifyTwo.numToBeAdded = 2
+	lister.removeEndpoint(0)
+	delete(expectedOne, "0")
+	delete(expectedTwo, "0")
+	notifyOne.numToBeRemoved = 1
+	notifyTwo.numToBeRemoved = 1
+
+	watcher.notifyOfEndpoints(notifyOne.ID(), notifyTwo.ID())
+
+	expectedOne["1"] = Endpoint{ID: "1"}
+	expectedOne["2"] = Endpoint{ID: "2"}
+	expectedOne["3"] = Endpoint{ID: "3"}
+	expectedTwo["2"] = Endpoint{ID: "2"}
+	expectedTwo["3"] = Endpoint{ID: "3"}
+
+	require.Equal(t, expectedOne, existingEndpoints(t, watcher, notifyOne.ID()))
+	require.Equal(t, expectedTwo, existingEndpoints(t, watcher, notifyTwo.ID()))
+
+	lister.updateEndpoint(2, "updated_target")
+	notifyOne.numToBeChanged = 1
+	notifyTwo.numToBeChanged = 1
+	watcher.notifyOfEndpoints(notifyOne.ID(), notifyTwo.ID())
+
+	expectedOne["2"] = Endpoint{ID: "2", Target: "updated_target"}
+	require.Equal(t, expectedOne, existingEndpoints(t, watcher, notifyOne.ID()))
+	expectedTwo["2"] = Endpoint{ID: "2", Target: "updated_target"}
+	require.Equal(t, expectedTwo, existingEndpoints(t, watcher, notifyTwo.ID()))
+
+	watcher.Unsubscribe(notifyOne)
+	require.Nil(t, existingEndpoints(t, watcher, notifyOne.ID()))
+
+	lister.addEndpoint(4)
+	notifyTwo.numToBeAdded = 1
+	watcher.ListAndWatch(notifyTwo)
+	expectedTwo["4"] = Endpoint{ID: "4"}
+	require.Equal(t, expectedTwo, existingEndpoints(t, watcher, notifyTwo.ID()))
+
+	watcher.Unsubscribe(notifyTwo)
+	require.Nil(t, existingEndpoints(t, watcher, notifyTwo.ID()))
+}
+
+func existingEndpoints(t *testing.T, watcher *EndpointsWatcher, id NotifyID) map[EndpointID]Endpoint {
+	if existing, ok := watcher.existingEndpoints.Load(id); ok {
+		endpoints, ok := existing.(map[EndpointID]Endpoint)
+		assert.True(t, ok)
+		return endpoints
 	}
+	return nil
+}
 
-	ew := EndpointsWatcher{
-		Endpointslister:   ml,
-		RefreshInterval:   2 * time.Second,
-		existingEndpoints: map[EndpointID]Endpoint{},
+func setup(t *testing.T) (*mockEndpointsLister, *EndpointsWatcher, *mockNotifier) {
+	ml := &mockEndpointsLister{
+		endpointsMap: map[EndpointID]Endpoint{},
 	}
 
-	mn := mockNotifier{}
+	ew := NewEndpointsWatcher(ml, 2*time.Second, zap.NewNop())
+	mn := newMockNotififier(t, "mockNotifier")
 
 	return ml, ew, mn
 }
 
+var _ Notify = (*mockNotifier)(nil)
+
 type mockNotifier struct {
```

Comment:
Potentially would be worthwhile using `mock.Mock` here, since we can make assertions on creation of the type and assert the number of expected calls.

Reply:
I've gone ahead and incorporated this, which make things a bit cleaner and more encompassing. Thanks!